asmcomp/arm64/emit.mlp

```diff
@@ -742,43 +754,47 @@ let emit_instr env i =
         assert (n mod 16 = 0);
         emit_stack_adjustment (-n);
         env.stack_offset <- env.stack_offset + n
-    | Lop(Iload(size, addr, _mut)) ->
+    | Lop(Iload { memory_chunk; addressing_mode; is_atomic }) ->
+        if is_atomic then ` dmb ishld\n`;
+        let load_type = if is_atomic then "ldar" else "ldr" in
         let dst = i.res.(0) in
         let base =
-          match addr with
+          match addressing_mode with
           | Iindexed _ -> i.arg.(0)
           | Ibased(s, ofs) ->
               assert (not !Clflags.dlcode);  (* see selection.ml *)
               `	adrp	{emit_reg reg_tmp1}, {emit_symbol_offset s ofs}\n`;
               reg_tmp1 in
-        begin match size with
+        begin match memory_chunk with
         | Byte_unsigned ->
-            `	ldrb	{emit_wreg dst}, {emit_addressing addr base}\n`
+            `	{emit_string load_type}b	{emit_wreg dst}, {emit_addressing addressing_mode base}\n`
         | Byte_signed ->
-            `	ldrsb	{emit_reg dst}, {emit_addressing addr base}\n`
+            `	{emit_string load_type}sb	{emit_reg dst}, {emit_addressing addressing_mode base}\n`
         | Sixteen_unsigned ->
-            `	ldrh	{emit_wreg dst}, {emit_addressing addr base}\n`
+            `	{emit_string load_type}h	{emit_wreg dst}, {emit_addressing addressing_mode base}\n`
         | Sixteen_signed ->
-            `	ldrsh	{emit_reg dst}, {emit_addressing addr base}\n`
+            `	{emit_string load_type}sh	{emit_reg dst}, {emit_addressing addressing_mode base}\n`
         | Thirtytwo_unsigned ->
-            `	ldr	{emit_wreg dst}, {emit_addressing addr base}\n`
+            `	{emit_string load_type}	{emit_wreg dst}, {emit_addressing addressing_mode base}\n`
         | Thirtytwo_signed ->
-            `	ldrsw	{emit_reg dst}, {emit_addressing addr base}\n`
+            `	{emit_string load_type}sw	{emit_reg dst}, {emit_addressing addressing_mode base}\n`
         | Single ->
-            `	ldr	s7, {emit_addressing addr base}\n`;
+            `	{emit_string load_type}	s7, {emit_addressing addressing_mode base}\n`;
             `	fcvt	{emit_reg dst}, s7\n`
         | Word_int | Word_val | Double ->
-            `	ldr	{emit_reg dst}, {emit_addressing addr base}\n`
+            `	{emit_string load_type}	{emit_reg dst}, {emit_addressing addressing_mode base}\n`
         end
-    | Lop(Istore(size, addr, _)) ->
+    | Lop(Istore(size, addr, initialization)) ->
```

Comment:
s/initialization/is_assigment? See

https://github.com/ocaml/ocaml/blob/9093386715688b5055b3b3258da73bfca2433613/asmcomp/mach.mli#L59-L60

`is_assignment` will be a better name for this variable.

Reply:
Maybe we should use a variant instead of a boolean if this causes bugs.

Reply:
Agreed.